share/CMakeJS.cmake

```diff
@@ -0,0 +1,642 @@
+    if(VERBOSE)
+        message(STATUS "NODE_EXECUTABLE: ${NODE_EXECUTABLE}")
+        message(STATUS "NODE_VERSION: ${NODE_VERSION}")
+    endif()
+endfunction()
+
+if(NOT DEFINED NODE_EXECUTABLE)
+    cmakejs_acquire_node_executable()
+    message(STATUS "NODE_EXECUTABLE: ${NODE_EXECUTABLE}")
+    message(STATUS "NODE_VERSION: ${NODE_VERSION}")
+endif()
+
+# Resolve NodeJS development headers
+# TODO: This code block is quite problematic, since:
+# 1 - it might trigger a build run, depending on how the builder has set up their package.json scripts...
+# 2 - it also currently assumes a preference for yarn over npm (and the others)...
+# 3 - finally, because of how cmake-js works, it might create Ninja-build artefacts,
+# even when the CMake user specifies a different generator to CMake manually...
+# We could use 'add_custom_target()' with a user-side ARG for which package manager to use...
+if(NOT IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/node_modules")
+    execute_process(
+      COMMAND yarn install
+      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+      OUTPUT_VARIABLE NODE_MODULES_DIR
+    )
+    if(NOT IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/node_modules")
+        message(FATAL_ERROR "Something went wrong - NodeJS modules installation failed!")
+        return()
+    endif()
+endif()
+
+#[=============================================================================[
+Get NodeJS C Addon development files.
+
+Provides
+::
+
+  NODE_API_HEADERS_DIR, where to find node_api.h, etc.
+  NODE_API_INC_FILES, the headers required to use Node Addon API.
+
+]=============================================================================]#
+function(cmakejs_acquire_napi_c_files)
+    execute_process(
+      COMMAND "${NODE_EXECUTABLE}" -p "require('node-api-headers').include_dir"
+      WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+      OUTPUT_VARIABLE NODE_API_HEADERS_DIR
+      # COMMAND_ERROR_IS_FATAL ANY
+    )
+    string(REGEX REPLACE "[\r\n\"]" "" NODE_API_HEADERS_DIR "${NODE_API_HEADERS_DIR}")
+    set(NODE_API_HEADERS_DIR "${NODE_API_HEADERS_DIR}" CACHE PATH "Node API Headers directory." FORCE)
+
+    unset(NODE_API_INC_FILES)
+    file(GLOB NODE_API_INC_FILES "${NODE_API_HEADERS_DIR}/*.h")
+    set(NODE_API_INC_FILES "${NODE_API_INC_FILES}" CACHE FILEPATH "Node API Header files." FORCE)
+    source_group("Node Addon API (C)" FILES "${NODE_API_INC_FILES}")
+
+    if(VERBOSE)
+        message(STATUS "NODE_API_HEADERS_DIR: ${NODE_API_HEADERS_DIR}")
+    endif()
+endfunction()
+
+# Acquire if needed...
+if(NOT DEFINED NODE_API_HEADERS_DIR)
+    cmakejs_acquire_napi_c_files()
+    message(STATUS "NODE_API_HEADERS_DIR: ${NODE_API_HEADERS_DIR}")
+    if(NOT DEFINED NODE_API_INC_FILES)
+        file(GLOB NODE_API_INC_FILES "${NODE_API_HEADERS_DIR}/*.h")
+        set(NODE_API_INC_FILES "${NODE_API_INC_FILES}" CACHE FILEPATH "Node API Header files." FORCE)
+        source_group("Node Addon API (C)" FILES "${NODE_API_INC_FILES}")
+    endif()
+endif()
```

Comment:
```suggestion
endfunction()

# Acquire if needed...
if(NOT DEFINED NODE_API_HEADERS_DIR)
    cmakejs_acquire_napi_c_files()
    message(STATUS "NODE_API_HEADERS_DIR: ${NODE_API_HEADERS_DIR}")
    unset(NODE_API_INC_FILES)
endif()
if(NOT DEFINED NODE_API_INC_FILES)
    file(GLOB NODE_API_INC_FILES "${NODE_API_HEADERS_DIR}/*.h")
    set(NODE_API_INC_FILES "${NODE_API_INC_FILES}" CACHE FILEPATH "Node API Header files." FORCE)
    source_group("Node Addon API (C)" FILES "${NODE_API_INC_FILES}")
endif()
```

I think this is more durable and avoids some duplication. It also lets the user override `NODE_API_HEADERS_DIR` without also needing to set `NODE_API_INC_FILES`

Reply:
> I think this is more durable and avoids some duplication. It also lets the user override `NODE_API_HEADERS_DIR` without also needing to set `NODE_API_INC_FILES`

Sorry for the sneaky change on this - the reason was because my targets as shown in VSCode's cmake-tools extension were showing some weird non-existent directory in their sources list... a file named 'dev' in location```../..```. 

![Screenshot from 2024-01-31 22-55-19](https://github.com/cmake-js/cmake-js/assets/97741588/1ea0dcf7-0c5d-4091-8568-9ed0cfc190ee)

Weirdly, the supposed file won't display, because VSCode says it's a directory (I am indeed working out of a directory with that name)...

These are those kinds of odd things that don't really show up until much later downstream, and then it is some huge de-tangling nightmare because of the brittle-ness of whatever CMake fun you've gotten yourself into...

On a related note, I highly recommend (any readers not already doing so) regularly doing a normal CMake configure on the test project and checking the content of these build dir artefacts:

- CMakeCache.txt
- share/cmake/CMakeJS**.cmake
- makefile or Ninja file or MSBuild files

In particular; paths, definitions, globbed files... "broken" stuff usually is quite easy to spot and diagnose in the output, whereas when you're just looking at the code you are putting in to CMake you might never know there was any issue.

Your suggestions and ideas are totally cool with me nonetheless. It just might be that a different solution is required, compared to my original draft...